src/concepts/tail.md

```diff
@@ -1 +1,93 @@
+If every recursive call made by a function is a **tail call**, that function is called **tail recursive**. Put another way, a function is **tail recursive** if the last operation performed by the function is the recursive call.
+
+Suppose we have the following function to sum the elements of a list:
+```sml
+fun sum [] = 0
+  | sum (x::xs) = x + (sum xs)
+```
+This function is not tail-recursive, because after the recursive call is evaluated, we add the result to `x`. Similarly, if we were to case on `sum xs` or did any other operation on its result, then this function would not be tail-recursive.
+
+In order to write the `sum` function in a tail-recursive manner, we know that the last computation that we can do is the recursive call to `sum`. But, we also need some way to keep track of the sum of the list elements and add to it, since we no longer can add things after the recursive call. We will make an accumulator variable, `acc`, that will keep track of the sum of list elements we have exposed so far, and pass that down through the recursive calls. In our base case, we know we have seen every element in the list and there are no more elements to add, so our base case looks like this:
+```sml
+fun sum' ([], acc) = acc
+```
+In our recursive case, we want to use our accumulator to account for the top element of our list, then pass down that accumulator to the recursive calls. We can do that as follows:
+```sml
+fun sum' (x::xs, acc) = sum' (xs, x + acc)
+```
+Because SML evaluates the function arguments before evaluating the function call, `x + acc` is performed before `sum` is called, and then this updated accumulator value is passed down to the recursive `sum` call on `xs`. Thus, the last operation performed is the recursive call, making this a tail-recursive function.
+
+Putting these two parts together, we have:
+
+```sml
+fun sum' ([], acc) = acc
+  | sum' (x::xs, acc) = sum' (xs, x + acc)
+```
+
+Why do we care about tail recursion? One reason is that the tail-recursive version of functions uses less space on the call stack.
+
+Consider the following stack trace of the `sum` function, which is not tail-recursive:
+```sml
+sum [3, 2, 1]
+=> 3 + (sum [2, 1])
+=> 3 + (2 + (sum [1]))
+=> 3 + (2 + (1 + (sum [])))
+=> 3 + (2 + (1 + (0)))
+=> 3 + (2 + (1))
+=> 3 + (3)
+=> 6
+```
+In this stack trace, we can see that the sum function takes linear space. (By stack space, we mean the space around the recursive call). Assuming an input list of length `n`, the stack will have `n` additions at its largest, giving us a stack of size `n`.
+
+Now, consider a stack trace of the tail-recursive `sum'` function:
+```sml
+sum ([3, 2, 1], 0)
+=> sum ([2, 1], 3)
+=> sum ([1], 5)
+=> sum ([], 6)
+=> 6
+```
+Notice that the stack trace does not get any wider: we do not need any memory space to store "what is left to do," so this takes constant space.
+
+# Example
+Let's try to turn the function to calculate the nth Fibonacci number into a tail-recursive function.
```

Comment:
not a change but i dont think ive ever thought of turning fib into a tail-recursive version - this is pretty neat!